### ISSUE TITLE
package level methods don't get compiled

### ISSUE THREAD

**Issue by rolandzwaga** [RandoriAS/randori-compiler#65]
If I add a package level method like this:

```
package mediators {
  public function test():void {
    var i:int = 0;
  }
}
```

no output is generated.

**Comment by teotigraphix** [comment]
Its not implemented as of now, only class compilation units. Will put it on the todo list.

**Comment by labriola** [comment]
More specifically, the function definition is output, however, the function's body is not. There is also some extra code generated which could be trimmed down as, as mike indicated, its not really a class

**Comment by labriola** [comment]
@teotigraphix , once we have that implemented, there are a few things in the compiler annotations doc that should also map onto package level functions as well. Also, please note, i removed one use case from the annotations document which is setting [JavaScriptMethod(mode="global")] This will be replaced by having the package level functions working

**Comment by teotigraphix** [comment]
@labriola So using the above code as an example, what is the exact java script code that should be generated and the .js file name?

**Comment by labriola** [comment]
if (typeof mediators == "undefined")
    var mediators = {};

mediators.test= function() {
  var i = 0;
};

mediators.test.className = "test";

mediators.test.getClassDependencies = function(t) {
    var p;
    return [];
};

mediators.test.injectionPoints = function(t) {
    return [];
};

**Comment by labriola** [comment]
the js file name would be test.js and in a mediators folder if using single file expansion

**Comment by labriola** [comment]
@teotigraphix the terminology above is code smell as we now have className for a function, getClassDependencies for the function's dependencies and there will never be injection points..... so, maybe in some far off refactor we might need to globally discuss names of those properties, but the code above should be good

**Comment by teotigraphix** [comment]
@labriola 

"now have className for a function, getClassDependencies for the function's dependencies and there will never be injection points....."

Are you talking about the footer generation? If so, I can easily omit that from a Function compilation unit.

**Comment by labriola** [comment]
We need to have it for things to class load. Just lamenting the names I chose :)

**Comment by labriola** [comment]
Mike, bonus round, if you get far enough take a look at how global interacts with package level functions in the annotation wiki

**Comment by teotigraphix** [comment]
I already have all this working. So my head is a bit fired, can you just explain in laymans terms what you are talking about?

Like I said, I have package level functions being generated now. We do need to talk about the footer emitter, I had to implement another one based off of Function Node, not ClassNode like the first one.

**Comment by rolandzwaga** [comment]
@teotigraphix have a look at the wiki page:
https://github.com/RandoriAS/randori-plugin-intellij/wiki/JavaScript-Compiler-Annotations
At the part below 'Package Level Function'
The comment 'Any attempted use of the package level function marked as global inside of ActionScript code should ideally raise an error.' is what Mike is refering to.
The instance where a package level method is annotated with mode="global" refers to the situation of unit testing. In that case, the methods marked as mode="global" are not meant to be called from anywhere inside of any other code. These methods will only be called by the testrunner. (The testrunner only looks for methods on the window instance.)
Therefore, it would be ideal if the compiler could raise an error should of these mtheods be called from anywhere in the source code.

Does that make sense?

**Comment by teotigraphix** [comment]
ok, makes sense.

So what you are saying is, the above example if annotated with a mode=global, the compiler would report a Problem because its usage is prohibited when visiting a FunctionCallNode.

**Comment by labriola** [comment]
So, just to be clear, if I have this:

package mediators {
  [JavaScript(mode="global")]
  public function test():void {
    var i:int = 0;
  }
}

The output should be:

var i:int = 0;

and if someone tries to call:

test();

the invocation of test() should throw an error as test() won't exist.

**Comment by teotigraphix** [comment]
ok, that is not what I was thinking but is clear now with the example.

Although it will probably end up like;

var i = 0;

The compiler doesn't like emitting types for javascript. :)

**Comment by labriola** [comment]
details ;)

**Comment by teotigraphix** [comment]
This will be working by Monday (export=global), package functions already work.

We need to discuss legal footer data IE injections. Class deps still get round up by types and the new operator in the function scope block, so that hasn't changed. And think about if you want to rename the className.

**Comment by labriola** [comment]
Injections can always look like this:

mediators.test.injectionPoints = function(t) {
return [];
};

It will never have injectionPoints but having the function there for now will make us handle one less case in the code

**Comment by teotigraphix** [comment]
ok, that is how it is right now, perfect.

This is what is being generated in the test right now;

```
if (typeof demo == "undefined")
    var demo = {};
if (typeof demo.foo == "undefined")
    demo.foo = {};

demo.foo.MyFunction = function() {
    var i = 0;
}

demo.foo.MyFunction.className = "demo.foo.MyFunction";

demo.foo.MyFunction.getClassDependencies = function(t) {
    var p;
    return [];
};

demo.foo.MyFunction.injectionPoints = function(t) {
    return [];
};
```

**Comment by labriola** [comment]
As long as MyFunction is the name of your function, it looks good

**Comment by teotigraphix** [comment]
This is the as;

```
package demo.foo
{
public function MyFunction():void {
    var i:int = 0;
}
}
```

**Comment by labriola** [comment]
Looks good to me. My comments on naming (e.g. className) are low priority and I dont think we will handle them soon. This looks great though

**Comment by teotigraphix** [comment]
Well, I'm done. :)

For this;

```
package demo.foo
{
[JavaScript(mode="global")]
public function MyTestFunction():void {
    var i:int = 0;
}
}
```

We now get;

```
var i = 0;
```

Great designs yield great efficiency.

**Comment by labriola** [comment]
Good, now get me global mode and our unit test runners with Karma will all work flawlessly

**Comment by teotigraphix** [comment]
What do you mean global mode? Isn't that what I just showed you I finished?

**Comment by labriola** [comment]
Sorry, was just saying I cant wait to have the build in hand. We are good

**Comment by teotigraphix** [comment]
I can get this today, one question. How is this global generated? All in one file, right now I am doing an individual file.

How should this work, am I missing something? Or are you just going to use the mono and exclude args I have in the compiler now?

**Comment by labriola** [comment]
Should follow the same rules as classes. So if we are exporting multi-file it would be a file per function. If we are in monolithic mode, they can be sorted with the other global stuff you do today.

**Comment by teotigraphix** [comment]
Ok that already works, just need to check if I am not casting something during the global filter.

**Comment by teotigraphix** [comment]
Unless anything else comes up, it needs to be a separate issue, this is closed

commit 979faa7
